CommunicationLayer/CommDeviceControl/UdpConnectionService.cpp

```diff
@@ -14,38 +26,83 @@ UdpConnectionService::~UdpConnectionService()
 
 void  UdpConnectionService::setMulticastNetwork(QHostAddress groupAddress, quint16 port)
 {
-    groupAddress_ = groupAddress;
-    port_ = port;
+    //groupAddress_ = groupAddress;
+    //port_ = port;
+
+
+
 }
 
-bool UdpConnectionService::connectToDataSource()
+void UdpConnectionService::setupChannel()
 {
-    if (groupAddress_.isNull() || port_ == 0)
-    {
-        emit connectionFailed("Group address and port number is not set!");
-        return false;
-    }
+    quint32 i = 0;
 
-    if (socket_.bind(QHostAddress::AnyIPv4, port_, QUdpSocket::ShareAddress) &&
-            socket_.joinMulticastGroup(groupAddress_))
-    {
-        emit connectionSucceeded();
-        return true;
-    }
-    else
+    do
     {
-        emit connectionFailed(socket_.errorString());
-        return false;
+        if (i++)
+        {
+            qWarning() << "UdpMessageForwarder: Attempting to reconnect" << ipAddress_ << udpPort_;
+        }
+
+        try
+        {
+            channel_ = AmqpClient::Channel::Create(ipAddress_.toStdString(), udpPort_);
+        }
+        catch (std::exception&)
+        {
+            if (channel_ == NULL)
+            {
+                if (i == (TIMEOUT / SLEEP_TIME))
+                {
+                    qWarning() << "UdpMessageForwarder timed out waiting for connection to broker";
+                    throw;
+                }
+
+                qWarning() << "UdpMessageForwarder: error creating channel";
+                QThread::sleep(SLEEP_TIME);
```

Comment:
Anything to do with threads is hard to debug, hard to code properly, and hard to test. QT has a great solution with signals and slots that should be taken advantage of. Look at QTimer if a task needs to be run periodically.